packages/prover/src/proof_provider/payload_store.ts

```diff
@@ -0,0 +1,154 @@
+
+  get finalized(): allForks.ExecutionPayload | undefined {
+    const finalizedMaxRoot = this.finalizedRoots.get(this.finalizedRoots.max);
+    if (finalizedMaxRoot) {
+      return this.payloads.get(finalizedMaxRoot);
+    }
+
+    return undefined;
+  }
+
+  get latest(): allForks.ExecutionPayload | undefined {
+    if (this.latestBlockRoot) {
+      return this.payloads.get(this.latestBlockRoot);
+    }
+
+    return undefined;
+  }
+
+  async get(blockId: number | string): Promise<allForks.ExecutionPayload | undefined> {
+    // Given block id is a block hash in hex (32 bytes root takes 64 hex chars + 2 for 0x prefix)
+    if (typeof blockId === "string" && blockId.startsWith("0x") && blockId.length === 64 + 2) {
+      return this.payloads.get(blockId);
+    }
+
+    // Given block id is a block number in hex
+    if (typeof blockId === "string" && blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(hexToNumber(blockId));
+    }
+
+    // Given block id is a block number in decimal string
+    if (typeof blockId === "string" && !blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(parseInt(blockId, 10));
+    }
+
+    // Given block id is a block number in decimal
+    if (typeof blockId === "number") {
+      return this.getOrFetchFinalizedPayload(blockId);
+    }
+
+    return undefined;
+  }
+
+  async getOrFetchFinalizedPayload(blockNumber: number): Promise<allForks.ExecutionPayload | undefined> {
+    if (blockNumber > this.finalizedRoots.max) {
+      throw new Error(
+        `Block number ${blockNumber} is higher than the latest finalized block number. We recommend to use block hash for unfinalized blocks.`
+      );
+    }
+
+    let blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (!blockELRoot) {
+      const payloads = await getExecutionPayloadForBlockNumber(this.opts.api, this.finalizedRoots.min, blockNumber);
+      for (const payload of Object.values(payloads)) {
+        this.set(payload, true);
+      }
+    }
+
+    blockELRoot = this.finalizedRoots.get(blockNumber);
+    if (blockELRoot) {
+      return this.payloads.get(blockELRoot);
+    }
+
+    return undefined;
+  }
+
+  set(payload: allForks.ExecutionPayload, finalized: boolean): void {
+    const blockRoot = bufferToHex(payload.blockHash);
+    this.payloads.set(blockRoot, payload);
+    this.latestBlockRoot = blockRoot;
+
+    if (finalized) {
+      this.finalizedRoots.set(payload.blockNumber, blockRoot);
+    }
+  }
+
+  async processLCHeader(header: capella.LightClientHeader, finalized = false): Promise<void> {
+    const blockSlot = header.beacon.slot;
+    const blockNumber = header.execution.blockNumber;
+    const blockELRoot = bufferToHex(header.execution.blockHash);
+    const blockCLRoot = bufferToHex(header.beacon.stateRoot);
+    const existingELRoot = this.unfinalizedRoots.get(blockCLRoot);
+
+    if (finalized) {
+      if (existingELRoot) {
+        this.unfinalizedRoots.delete(blockCLRoot);
```

Comment:
we might need to delete all `unfinalizedRoots` which have slot <= finalized

Reply:
We don't store the `slot` numbers for the `unfinalizedRoots` so can't delete those with that condition. I added an action item to revisit that logic for future. Once we add extended unit tests this logic will be easy to verify and refactor.

Reply:
but you have payloads stored for them, so do a look and cleanup